### DISCUSSION TITLE
Formatter: append multiple lines seems not to work as expected (prepend works) - resolved: I accidentally have overseen a conflict

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1529] [Q&A] [answered]
I have a Grammar (languim 3.1) as follows (modified hello-world):
```
entry Model:
    (groups+=Group)+;

Group:
    'group' name=ID '{'
        (persons+=Person | greetings+=Greeting | groups+=Group)*
     '}';

Person:
    'person' name=ID info=STRING;

Greeting:
    'Hello' person=[Person:ID] '!';
```

with a formatter as follows:

```javascript
    protected override format(node: AstNode): void {
        if (ast.isGroup(node)) {
            const formatter = this.getNodeFormatter(node);
            const start = formatter.keyword('group');
            const bracesOpen = formatter.keyword('{');
            const bracesClose = formatter.keyword('}');
            start.prepend(Formatting.newLines(3));
            formatter.interior(bracesOpen, bracesClose).prepend(Formatting.indent());
            bracesOpen.prepend(Formatting.newLine());
            bracesClose.prepend(Formatting.newLine());
            formatter.property('name').surround(Formatting.oneSpace());
        } else if (ast.isModel(node)) {
            const formatter = this.getNodeFormatter(node);
            const nodes = formatter.nodes(...node.groups);
            nodes.prepend(Formatting.noIndent());
        }
    }
```

This works as expected: a model is formatted like in the following example:
```



group hello
{
  person Tom "not jerry"
  person Jerry "not a cat...a mouse"


  group inner
  {
    person Tim "not a dog"
    person Struppi "not a human...an animal"
  }
}
```

However, when changing `prepend` to `append` (`start.append(Formatting.newLines(3));`) no empty lines are inserted anywhere... Is this related to some open issue or did I misunderstand the described function call?

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

at least in this case, you run into conflicting formattings. On one hand you have `formatter.property('name').surround(Formatting.oneSpace())`, which expects that between `group hello` there is exactly one space. On the other hand you have `start.append(Formatting.newLines(3))`, which expects that between `group hello` there are three new lines. Since the `surround(Formatting.oneSpace())` call comes last, it receives precedence.

**Reply by goto40** [reply]
Thank you for the quick reply!

**Reply by msujew** [reply]
IIRC you can specify a priority on the formattings (the `Formatting` functions have an optional parameter for that). A higher priority gets precedence in case of conflicts.

**Reply by goto40** [reply]
Thanks again. However I probably misused the priority somehow... In the following case the `property('name').surround` wins in all cases. Only when commenting out the `name`-formatting I get an effect from `start.append`:

```javascript
        if (ast.isGroup(node)) {
            const formatter = this.getNodeFormatter(node);
            const start = formatter.keyword('group');
            const bracesOpen = formatter.keyword('{');
            const bracesClose = formatter.keyword('}');
            start.append(Formatting.newLines(3,{priority: 10})); // this one has no effect (except I comment the line **1**)
            formatter.interior(bracesOpen, bracesClose).prepend(Formatting.indent());
            bracesOpen.prepend(Formatting.newLine());
            bracesClose.prepend(Formatting.newLine());
            formatter.property('name').surround(Formatting.oneSpace({priority: 0})); // this one (**1**) wins
        } ...
```

**Reply by msujew** [reply]
Hm, might be that the priority based conflict resolution doesn't work correctly on that case. That another bug/issue I need to track for the formatter improvements I'm planning to do soon (probably for a 4.0 - it's likely minorly API breaking).
